lib/puppetserver/ca/certificate_authority.rb

```diff
@@ -0,0 +1,210 @@
+require 'puppetserver/ca/utils/http_client'
+
+require 'json'
+
+module Puppetserver
+  module Ca
+    class CertificateAuthority
+
+      include Puppetserver::Ca::Utils
+
+      REVOKE_BODY = JSON.dump({ desired_state: 'revoked' })
+      SIGN_BODY =   JSON.dump({ desired_state: 'signed' })
+
+      def initialize(logger, settings)
+        @logger = logger
+        @client = HttpClient.new(settings)
+        @ca_server = settings[:ca_server]
+        @ca_port = settings[:ca_port]
+      end
+
+      # Returns a URI-like wrapper around CA specific urls
+      def make_ca_url(resource_type = nil, certname = nil)
+        HttpClient::URL.new('https', @ca_server, @ca_port, 'puppet-ca', 'v1', resource_type, certname)
+      end
+
+      def sign_certs(certnames)
+        put('certificate_status', certnames, SIGN_BODY, :sign)
+      end
+
+      def revoke_certs(certnames)
+        put('certificate_status', certnames, REVOKE_BODY, :revoke)
+      end
+
+      # Should this csr be in JSON? instead of string?
```

Comment:
I'm sending the csr as a string, which works. Should it be JSON.dump like the other bodies?

Reply:
Should it be JSON? it looks like it csr is an OpenSSL object? I think then `to_s` is alaised to  `to_pem`? The docs say that it should be sent as a `Content-Type` `plain/text` but contain just the pem encoded CSR: https://puppet.com/docs/puppet/5.5/http_api/http_certificate_request.html#save

Reply:
We definitely use `text/plain` in the agent.

Reply:
I'm updated it to be .to_pem for clarity and have this call use custom headers with content-type text/plain